d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -239,6 +300,68 @@ XdsClient.D2NodeMapResourceWatcher getUriResourceWatcher(String clusterName)
     return new UriPropertiesResourceWatcher(clusterName);
   }
 
+  XdsClient.D2SymlinkNodeResourceWatcher getSymlinkResourceWatcher(String symlinkName)
+  {
+    return new XdsClient.D2SymlinkNodeResourceWatcher()
+    {
+      @Override
+      public void onChanged(String resourceName, XdsClient.D2SymlinkNodeUpdate update)
+      {
+        // update maps between symlink name and actual node name, listen to the actual node
+        String actualResourceName = update.getNodeData().getMasterClusterNodePath();
+        if (resourceName.contains(D2_CLUSTER_NODE_PREFIX))
+        {
+          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_CLUSTER_NODE_PREFIX);
+          updateSymlinkAndActualNodeMap(symlinkName, actualNodeName);
+          listenToCluster(actualNodeName);
+        }
+        else
+        {
+          String actualNodeName = removeNodePathPrefix(actualResourceName, D2_URI_NODE_PREFIX);
+          updateSymlinkAndActualNodeMap(symlinkName, actualNodeName);
+          listenToUris(actualNodeName);
+        }
+      }
+
+      @Override
+      public void onError(Status error)
+      {
+        notifyAvailabilityChanges(false);
+      }
+
+      @Override
+      public void onReconnect()
+      {
+        notifyAvailabilityChanges(true);
+      }
+    };
+  }
+
+  private void updateSymlinkAndActualNodeMap(String symlinkName, String actualNodeName) {
+    synchronized (_symlinkAndActualNodeLock) {
+      _symlinkAndActualNode.put(symlinkName, actualNodeName);
+    }
+  }
+
+  private String getSymlink(String actualNodeName) {
+    synchronized (_symlinkAndActualNodeLock) {
+      return _symlinkAndActualNode.inverse().get(actualNodeName);
+    }
+  }
+
+  private static String removeNodePathPrefix(String path, String prefix)
```

Comment:
Can `SymlinkUtil` be leveraged here?

Reply:
Here we need to find the index of "$" and remove all preceding path. SymlinkUtil finds the end of the symlink:
 /**
   * return the index to split the first symlink out of param path.
   *
   * If no symlink is found, -1 is returned.
   *
   * Examples:
   *  firstSymlinkIndex("/$symlink/foo/bar") returns 9
   *  firstSymlinkIndex("/foo/bar/$symlink") returns 17
   *  firstSymlinkIndex("/notASymlink") returns -1
   *